src/utils/libftdi_serial.cpp

```diff
@@ -0,0 +1,252 @@
+            return (int)SerialExitCodes::OK;
+        case -3: // usb device unavailable
+            log_error ("set_custom_baudrate");
+            return (int)SerialExitCodes::OPEN_PORT_ERROR;
+        default: // -1 invalid baudrate, -2 setting baudrate failed
+            log_error ("set_custom_baudrate");
+            return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+    }
+}
+
+int LibFTDISerial::flush_buffer ()
+{
+#if FTDI_MAJOR_VERSION > 2 || (FTDI_MAJOR_VERSION == 1 && FTDI_MINOR_VERSIOM >= 5)
+    // correct tcflush was added in libftdi 1.5
+    switch (ftdi_tcioflush (ftdi))
+    {
+        case 0:
+            return (int)SerialExitCodes::OK;
+        case -3: // usb device unavailable
+            log_error ("flush_buffer ()");
+            return (int)SerialExitCodes::OPEN_PORT_ERROR;
+        default: // -1,-2 chip failed to purge a buffer
+            log_error ("flush_buffer ()");
+    }
+#else
+    log_error ("flush_buffer ()", "libftdi version <=1.4, tcflush unimplemented");
+#endif
+    return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+}
+
+int LibFTDISerial::read_from_serial_port (void *bytes_to_read, int size)
+{
+    // the ftdi will send us data after its latency (max 255ms, default
+    // 16ms) times out, even if its buffer is empty, despite the usb
+    // timeout.  libftdi does not enforce waiting for the usb timeout if
+    // the chip responds, even if the chip responds with an empty buffer.
+    // so, the read is repeated until the timeout is reached.
+
+    // this latency behavior is documented in
+    // http://www.ftdichip.com/Support/Documents/AppNotes/AN232B-04_DataLatencyFlow.pdf
+
+    auto deadline =
+        std::chrono::steady_clock::now () + std::chrono::milliseconds (ftdi->usb_read_timeout);
+    int bytes_read = 0;
+    while (bytes_read == 0 && size > 0 && std::chrono::steady_clock::now () < deadline)
+    {
+        bytes_read = ftdi_read_data (ftdi, static_cast<unsigned char *> (bytes_to_read), size);
+        // TODO: negative values are libusb error codes, -666 means usb device unavailable
+        if (bytes_read < 0)
+        {
+            log_error ("read_from_serial_port");
+        }
+    }
+
+    return bytes_read;
+}
+
+int LibFTDISerial::send_to_serial_port (const void *message, int length)
+{
+    int bytes_written =
+        ftdi_write_data (ftdi, static_cast<unsigned const char *> (message), length);
+    // TODO: negative values are libusb error codes, -666 means usb device unavailable
+    if (bytes_written < 0)
+    {
+        log_error ("send_to_serial_port");
+    }
+    return bytes_written;
+}
+
+int LibFTDISerial::close_serial_port ()
+{
+    if (ftdi_usb_close (ftdi) == 0)
+    {
+        port_open = false;
+        return (int)SerialExitCodes::OK;
+    }
+    else
+    {
+        log_error ("close_serial_port ()");
+        return (int)SerialExitCodes::CLOSE_ERROR;
+    }
+}
+
+const char *LibFTDISerial::get_port_name ()
+{
+    return description.c_str ();
+}
+
+#else
```

Comment:
I might be missing something, but it looks to me like this block is unneeded if the libftdi interface is kept private, because the only way to access it is guarded by the same check: https://github.com/brainflow-dev/brainflow/pull/221/files#diff-707d50ff0c1f53ceed4b00bbb44de221f4e4804adeb6c8a62059677a2f153933L443

Reply:
no, because now there is a header for libftdi class and this header included wo ifdef. Can change it and include header under ifdef check also, I like this way because there is no need to write ifdefs outside class implementation. But its just a preference, both options are fine